### PR TITLE
refactor PBFTEngine for code reuse

### PR DESCRIPTION
**Pull request by cyjseagull**
This change is [Reviewable](https://reviewable.io/reviews/fisco-bcos/fisco-bcos/827)

### CODE REVIEW

File: libconsensus/pbft/PBFTEngine.cpp [hunk 1]
```diff
@@ -919,96 +921,102 @@ void PBFTEngine::checkAndCommit()
...
+    }
+}
+
+/// if collect >= 2/3 SignReq and CommitReq, then callback this function to commit block
+/// check whether view and height is valid, if valid, then commit the block and clear the context
+void PBFTEngine::checkAndSave()
+{
+    size_t sign_size = m_reqCache->getSigCacheSize(m_reqCache->prepareCache()->block_hash);
+    size_t commit_size = m_reqCache->getCommitCacheSize(m_reqCache->prepareCache()->block_hash);
+    if (sign_size >= minValidNodes() && commit_size >= minValidNodes())
```
Comment (wheatli) [inline libconsensus/pbft/PBFTEngine.cpp]: do not call minValidNdoes() twice~
```suggestion
    auto min_nodes = minValidNodes()
    if (sign_size >= min_nodes && commit_size >= min_nodes)
```

Comment (cyjseagull) [reply libconsensus/pbft/PBFTEngine.cpp]: done, pls refer to: https://github.com/FISCO-BCOS/FISCO-BCOS/pull/827/files#diff-b7e4695daa9e7f6fd6da2fc48711cd6cR1017

File: libconsensus/pbft/PBFTBroadcastCache.h [hunk 2]
```diff
@@ -0,0 +1,88 @@
...
+     * @param nodeId : node id
+     * @param type: packet type
+     * @param key: key (mainly the signature of specified broadcast packet)
+     * @return true: insert success
+     * @return false: insert failed
+     */
+    inline bool insertKey(h512 const& nodeId, unsigned const& type, std::string const& key)
+    {
+        if (!m_broadCastKeyCache.count(nodeId))
+            m_broadCastKeyCache[nodeId] = m_pbftReqFactory->buildPBFTMsgCache();
```
Comment (bxq2011hust) [inline libconsensus/pbft/PBFTBroadcastCache.h]:
```suggestion
 if (!m_broadCastKeyCache.count(nodeId))
 {
            m_broadCastKeyCache[nodeId] = m_pbftReqFactory->buildPBFTMsgCache();
  }
```

File: libconsensus/pbft/PBFTBroadcastCache.h [hunk 3]
```diff
@@ -0,0 +1,88 @@
...
+    {
+        if (!m_broadCastKeyCache.count(nodeId))
+            return false;
+        return m_broadCastKeyCache[nodeId]->exists(type, key);
+    }
+
+    /// clear all caches
+    inline void clearAll()
+    {
+        for (auto& item : m_broadCastKeyCache)
```
Comment (bxq2011hust) [inline libconsensus/pbft/PBFTBroadcastCache.h]:
```suggestion
        for (auto& item : m_broadCastKeyCache)
        {
            item.second->clearAll();
        }
```

File: libconsensus/pbft/PBFTBroadcastCache.h (hunk 2 above)
Comment (cyjseagull) [reply libconsensus/pbft/PBFTBroadcastCache.h]: modified,  already add brackets around all the `if` conditions.

File: libconsensus/pbft/PBFTBroadcastCache.h (hunk 3 above)
Comment (cyjseagull) [reply libconsensus/pbft/PBFTBroadcastCache.h]: modified,  already add brackets around all the `if` conditions.

**Pull request merged**